src/_P037_MQTTImport.ino

```diff
@@ -138,66 +223,320 @@ boolean Plugin_037(byte function, struct EventStruct *event, String& string)
+                  floatPayload = NAN;     // Invalid value
+                }
+                UserVar[event->BaseVarIndex + x] = floatPayload;							// Save the new value
+
+                if (!checkJson && P037_SEND_EVENTS && Settings.UseRules) { // Generate event of all non-json topic/payloads
+                  String RuleEvent;
+                  RuleEvent  = getTaskDeviceName(event->TaskIndex);
+                  RuleEvent += '#';
+                  RuleEvent += event->String1;
+                  RuleEvent += '=';
+                  RuleEvent += unparsedPayload;
+                  eventQueue.add(RuleEvent);
+                }
+
+                // Log the event
+                if (loglevelActiveFor(LOG_LEVEL_INFO)) {
+                  String log = F("IMPT : [");
+                  log += getTaskDeviceName(event->TaskIndex);
+                  log += '#';
+                  if (checkJson) {
+                    log += key;
+                  } else {
+                    log += ExtraTaskSettings.TaskDeviceValueNames[x];
+                  }
+                  log += F("] : ");
+                  log += floatPayload;
+                  addLog(LOG_LEVEL_INFO, log);
+                }
+
+                // Generate event for rules processing - proposed by TridentTD
+
+                if (Settings.UseRules) {
+                  String RuleEvent;
+                  if (checkJson) {
+                    // For JSON payloads generate <Topic>#<Attribute>=<Payload> event
+                    RuleEvent  = event->String1;
+#if defined(P037_FILTER_SUPPORT) && defined(P037_FILTER_PER_TOPIC)
+                    RuleEvent += P037_data->getFilterAsTopic(x + 1);
+#endif
+                    RuleEvent += '#';
+                    RuleEvent += key;
+                    RuleEvent += '=';
+                    bool hasSemicolon = unparsedPayload.indexOf(';') > -1;
+                    if (numericPayload && !hasSemicolon) {
+                      RuleEvent += floatPayload;
+                    } else if (numericPayload && hasSemicolon) { // semicolon separated list, pass unparsed
+                      RuleEvent += Payload;
+                      RuleEvent += ',';
+                      RuleEvent += unparsedPayload;
+                    } else {
+                      RuleEvent += Payload; // Pass mapped result
+                    }
+                    eventQueue.add(RuleEvent);
+                  } 
+                  // (Always) Generate <Taskname>#<Valuename>=<Payload> event
+                  RuleEvent  = getTaskDeviceName(event->TaskIndex);
+                  RuleEvent += '#';
+                  RuleEvent += ExtraTaskSettings.TaskDeviceValueNames[x];
+                  RuleEvent += '=';
+                  if (numericPayload) {
+                    RuleEvent += floatPayload;
+                  } else {
+                    RuleEvent += Payload;
+                  }
+                  eventQueue.add(RuleEvent);
```

Comment:
Maybe consider combining events as you may trigger a lot of events.

Reply:
Combining is quite hard, as the number of Json attributes is unknown until after processing them. I'll try to think of a solution here, but it should be 'solved' by using filtering.